src/util/getNextThoughtsWithContextChain.js

```diff
@@ -0,0 +1,178 @@
+import { store } from '../store.js'
+
+import {
+  RANKED_ROOT,
+  ROOT_TOKEN,
+} from '../constants.js'
+// util
+import {
+  contextOf,
+  getContextsSortedAndRanked,
+  getSortPreference,
+  getThought,
+  getThoughtsRanked,
+  getThoughtsSorted,
+  head,
+  isFunction,
+  meta,
+  pathToContext,
+  splitChain,
+  nextSibling as thoughtNextSibling,
+  unroot,
+} from '../util.js'
+import { isContextViewActive } from './isContextViewActive.js'
+
+const contextToThought = (context, thoughtIndex) => context.map(contextValue => {
+  const thought = getThought(contextValue, thoughtIndex)
+  return { value: thought.value, rank: thought.contexts[0] && thought.contexts[0].rank }
+})
+
+const mapContextsToThoughts = (contexts, thoughtIndex) => contexts.map(({ context }) => contextToThought(context, thoughtIndex))
+
+const isValidContextView = context => {
+  // const allowSingleContext = false
+  if (!context.length) {
+    return false
+  }
+  return isContextViewActive(context) && getContextsSortedAndRanked(head(pathToContext(context))).length > 1
+}
+
+const getContextSiblings = (context, thoughtIndex) => {
+  const contexts = getContextsSortedAndRanked(head(context))
+  return mapContextsToThoughts(contexts, thoughtIndex)
+}
+
+const nextSiblingContext = (value, rank, context, thoughtIndex) => {
+  const contextSiblings = getContextSiblings(context, thoughtIndex)
+  const currentIndex = contextSiblings.findIndex(thoughts => thoughts.find(thought => thought.value === value))
+  return contextSiblings[currentIndex + 1]
+}
+
+const firstChildOfContext = (value, cursor, thoughtIndex) => {
+  const context = pathToContext(cursor)
+  const contextSiblings = getContextSiblings(context, thoughtIndex)
+  return contextSiblings[0]
+}
+
+const firstChildOfThought = (thoughtsRanked, showHiddenThoughts) => {
+  const contextMeta = meta(thoughtsRanked)
+  const sortPreference = getSortPreference(contextMeta)
+  const children = (sortPreference === 'Alphabetical' ? getThoughtsSorted : getThoughtsRanked)(thoughtsRanked)
+  const notHidden = child => !isFunction(child.value) && !meta(pathToContext(thoughtsRanked).concat(child.value)).hidden
+  const childrenFiltered = showHiddenThoughts ? children : children.filter(notHidden)
+  return childrenFiltered[0]
+}
+
+const getDerivedCursor = contextChain => {
```

Comment:
What is a "derived cursor"? When introducing a new term not used anywhere before in the codebase, please include a description.

Reply:
I've updated the terminologies to make them sound more explicit.